### PR TITLE
Add Java and Python Solution for Distributing Apples (cses/1716)

### PR DESCRIPTION
**Pull request by ChuangSheep**
_Place an "x" in the corresponding checkbox if it is done or does not apply to this pull request._

- [x] I have tested my code.
- [x] I have added my solution according to the steps [here](https://usaco.guide/general/adding-solution#steps).
- [x] I have followed the code conventions mentioned [here](https://usaco.guide/general/adding-solution/#code-conventions).
  - I understand that if it is clear that I have not attempted to follow these conventions, my PR will be closed.
  - If changes are requested, I will re-request a review after addressing them.
- [x] I have linked this PR to any issues that it closes.

### CODE REVIEW

**Review by dongliu0426 — APPROVED** [review]

**Comment by ChuangSheep** [comment]
image
image

"nEw pOliCy"

**Comment by SansPapyrus683** [comment]
it's to avoid another #3973
unless you have a better way that's how i think we should do it

File: solutions/gold/cses-1716.mdx [hunk 1]
```diff
@@ -97,6 +99,95 @@ int main() {
...
+import java.util.*;
+
+public class Main {
+	static final long MOD = (long)1e9 + 7;
+	static final int MAXN = (int)2e6;
+
+	static long[] fac = new long[MAXN];
+	static long[] inv = new long[MAXN];
+
+	static long binpow(long x, long n, long m) {
```
Comment (SansPapyrus683) [inline solutions/gold/cses-1716.mdx:116]: quick doc comment w/ source?

File: solutions/gold/cses-1716.mdx [hunk 2]
```diff
@@ -97,6 +99,94 @@ int main() {
...
+
+	public static void main(String[] args) throws IOException {
+		fac[0] = fac[1] = 1;
+		for (int i = 2; i < MAXN; i++) { fac[i] = i * fac[i - 1] % MOD; }
+
+		inv[0] = 1;
+		inv[MAXN - 1] = binpow(fac[MAXN - 1], MOD - 2, MOD);
+		for (int i = MAXN - 2; i > 0; i--) {
+			inv[i] = (i + 1) * inv[i + 1] % MOD;
+		}
```
Comment (ryanchou-dev) [inline solutions/gold/cses-1716.mdx:139]: add a comment here about computing inverses

**Review by ryanchou-dev — COMMENTED** [review]
maybe add a couple comments but looks good 👍

File: solutions/gold/cses-1716.mdx [hunk 3]
```diff
@@ -97,6 +99,95 @@ int main() {
...
+
+```py
+MOD: int = int(1e9) + 7
+MAXN: int = int(2e6)
+
+fac: list = [1 for _ in range(MAXN)]
+inv: list = [1 for _ in range(MAXN)]
+
+
+def binpow(x: int, n: int, m: int) -> int:
```
[the remark this answers is not in the document]
Comment (ChuangSheep) [reply solutions/gold/cses-1716.mdx:174]: na just for consistency

**Review by SansPapyrus683 — APPROVED** [review]

**Pull request merged**